spacy/tests/doc/test_json_doc_conversion.py

```diff
@@ -57,18 +99,116 @@ def test_doc_to_json(doc):
+        Token.set_extension("token_test", default=False)
+    if not Span.has_extension("span_test"):
+        Span.set_extension("span_test", default=False)
+
+    doc._.json_test1 = "hello world"
+    doc._.json_test2 = [1, 2, 3]
+    doc[0:1]._.span_test = "span_attribute"
+    doc[0]._.token_test = 117
+    doc.spans["span_group"] = [doc[0:1]]
+    json_doc = doc.to_json(
+        underscore=["json_test1", "json_test2", "token_test", "span_test"]
+    )
+
+    assert "_" in json_doc
+    assert json_doc["_"]["json_test1"] == "hello world"
+    assert json_doc["_"]["json_test2"] == [1, 2, 3]
+    assert "user_data" in json_doc
+    assert json_doc["user_data"]["tokens"]["token_test"]["value"] == 117
+    assert json_doc["user_data"]["spans"]["span_test"]["value"] == "span_attribute"
+    assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
+
+
+def test_doc_to_json_with_custom_user_data(doc):
+    if not Doc.has_extension("json_test"):
+        Doc.set_extension("json_test", default=False)
+    if not Token.has_extension("token_test"):
+        Token.set_extension("token_test", default=False)
+    if not Span.has_extension("span_test"):
+        Span.set_extension("span_test", default=False)
+
+    doc._.json_test = "hello world"
+    doc[0:1]._.span_test = "span_attribute"
+    doc[0]._.token_test = 117
+    json_doc = doc.to_json(underscore=["json_test", "token_test", "span_test"])
+    doc.user_data["user_data_test"] = 10
+    doc.user_data[("user_data_test2", True)] = 10
+
+    assert "_" in json_doc
+    assert json_doc["_"]["json_test"] == "hello world"
+    assert "user_data" in json_doc
+    assert json_doc["user_data"]["tokens"]["token_test"]["value"] == 117
+    assert json_doc["user_data"]["spans"]["span_test"]["value"] == "span_attribute"
+    assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
+
+
+def test_doc_to_json_with_token_span_same_identifier(doc):
+    if not Doc.has_extension("my_ext"):
+        Doc.set_extension("my_ext", default=False)
+    if not Doc.has_extension("my_ext"):
+        Doc.set_extension("my_ext", default=False)
+    if not Token.has_extension("my_ext"):
+        Token.set_extension("my_ext", default=False)
+    if not Span.has_extension("my_ext"):
+        Span.set_extension("my_ext", default=False)
+
+    doc._.my_ext = "hello world"
+    doc[0:1]._.my_ext = "span_attribute"
+    doc[0]._.my_ext = 117
+    json_doc = doc.to_json(underscore=["my_ext"])
+
+    assert "_" in json_doc
+    assert json_doc["_"]["my_ext"] == "hello world"
+    assert "user_data" in json_doc
+    assert json_doc["user_data"]["tokens"]["my_ext"]["value"] == 117
+    assert json_doc["user_data"]["spans"]["my_ext"]["value"] == "span_attribute"
+    assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
```

Comment:
For all the tests that include all the available extensions, I'd suggest adding a full doc round-trip that checks `doc.to_bytes() == new_doc.to_bytes`.

Reply:
This actually throws assertion errors

Reply:
I also added the assert line to the `test_json_to_doc` which doesn't include any custom attributes. Would this indicate that something is wrong with `to_json()` / `from_json()` or rather `to_bytes()` ?

Reply:
Can you figure out what the differences are between the docs?

Reply:
As far as I can tell, this seems to be the only difference between the two docs.

```
E         -  b'\xa1d\xa3ORG\xa4ROOT\xa7Feat1=B\xb2has_unknown_spaces\xc2',
E         ?                                                            ^
E         +  b'\xa1d\xa3ORG\xa4ROOT\xa7Feat1=B\xb2has_unknown_spaces\xc3',
E         ?                                                            ^
```

Reply:
The end byte of the doc seems to differ, maybe related to encoding?

Reply:
I can't tell which one, but it looks like one of them was created with `spaces` and one without. I'm not sure if it's a problem in the test doc/fixture or in `Doc.from_json`.

`Doc.to_json` doesn't really handle `has_unknown_spaces=True`, though, so it could be a more general problem.

Reply:
It's not the encoding, it's `has_unknown_spaces`. I think we can pretty much that we only want to handle `has_unknown_spaces=False`, but I'm not sure how far we want to enforce it. For the tests it should be possible create input/output docs that both have `has_unknown_spaces=False`.

Reply:
I got it fixed by setting both `has_unknown_spaces=False`. However, there is still a different assertion error when comparing both `to_bytes()`

```
E         -  b'r_data_values\xc4!\x94\xabhello world\x93\x01\x02\x03u\xaespan_attribute',
E         ?                                                         -
E         +  b'r_data_values\xc4!\x94\xabhello world\x93\x01\x02\x03\xaespan_attributeu',
E         ? 
```      

It has to do something with the custom span attribute, the `new_doc` seems to add an `u` at the end? I haven't figured out yet what it means / how it is caused.